src/validate.rs

```diff
@@ -0,0 +1,160 @@
+}
+
+pub(crate) fn validate_receive_individual(
+    mapping: MappingRef<&Interface>,
+    data: &AstarteType,
+) -> Result<(), ValidationError> {
+    if data.is_unset() && !mapping.allow_unset() {
+        return Err(ValidationError::Unset);
+    }
+
+    Ok(())
+}
+
+pub(crate) fn validate_receive_object(
+    aggregate: &HashMap<String, AstarteType>,
+) -> Result<(), ValidationError> {
+    for value in aggregate.values() {
+        if value.is_unset() {
+            return Err(ValidationError::Unset);
+        }
+    }
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use crate::{interfaces::tests::DEVICE_OBJECT, mapping};
+
+    use super::*;
+
+    use chrono::{TimeZone, Utc};
+
+    const DEVICE_DATASTREAM: &str = include_str!(
+        "../tests/e2etest/interfaces/org.astarte-platform.rust.e2etest.DeviceDatastream.json"
+    );
+
+    fn inititialize_aggregate() -> (Interface, HashMap<String, AstarteType>) {
+        let aggregate = HashMap::from_iter([
+            (
+                "double_endpoint".to_string(),
+                AstarteType::Double(37.534543),
+            ),
+            ("integer_endpoint".to_string(), AstarteType::Integer(45)),
+            ("boolean_endpoint".to_string(), AstarteType::Boolean(true)),
+            (
+                "booleanarray_endpoint".to_string(),
+                AstarteType::BooleanArray(vec![true, false, true]),
+            ),
+        ]);
+
+        let interface = Interface::from_str(DEVICE_OBJECT).unwrap();
+
+        (interface, aggregate)
+    }
+
+    #[test]
+    fn test_validate_send_for_aggregate_datastream() {
+        let (interface, _aggregate) = inititialize_aggregate();
+        let object = interface.as_object_ref().unwrap();
+
+        // Test sending an aggregate (with and without timestamp)
+        let timestamp = Some(TimeZone::timestamp_opt(&Utc, 1537449422, 0).unwrap());
+        validate_send_object(object, &None).unwrap();
+        validate_send_object(object, &timestamp).unwrap();
+    }
+
+    #[test]
+    fn test_validate_send_for_aggregate_datastream_extra_field() {
+        let (interface, mut aggregate) = inititialize_aggregate();
+        let object = interface.as_object_ref().unwrap();
+
+        // Test sending an aggregate with an non existing object field
+        aggregate.insert("gibberish".to_string(), AstarteType::Boolean(false));
+        let res = validate_send_object(object, &None);
+        assert!(res.is_ok());
+    }
+
+    #[test]
+    fn test_validate_send_for_individual_datastream() {
+        let interface = Interface::from_str(DEVICE_DATASTREAM).unwrap();
+
+        let mapping = MappingRef::new(&interface, mapping!("/boolean_endpoint")).unwrap();
+
+        let res = validate_send_individual(mapping, &AstarteType::Boolean(true), &None);
+        assert!(res.is_ok());
+
+        let res = validate_send_individual(mapping, &AstarteType::Boolean(true), &Some(Utc::now()));
+        assert!(res.is_ok());
+    }
```

Comment:
Add validation tests for sending set/unset properties and for receiving values.

Reply:
I will add the test for the unset in the deserialization function, since the check has been moved there.

Reply:
Done